File: java/KClosestPointsToOrigin.java
class Solution {
    // https://leetcode.com/problems/k-closest-points-to-origin/solution/
    private double getEuclideanDistance(int[] xy) {
        // Euclidean distance: c = ((xA − xB)^2 + (yA − yB)^2)^(1/2)
        return Math.sqrt(xy[0]*xy[0] + xy[1]*xy[1]); 
    }
    public int[][] kClosest(int[][] points, int K) {
        PriorityQueue<int[]> closestPoints = new PriorityQueue<>(K, new Comparator<int[]>() {
           @Override
            public int compare(int[] a, int[] b) {
                return Double.compare(getEuclideanDistance(a), getEuclideanDistance(b));
            }
        });
        
        
        for(int[] point: points) {
            closestPoints.offer(point);
        }
        
        int[][] result = new int[K][2];
        for(int i=0; i<K; ++i) {
            result[i] = closestPoints.poll();
        }
        
        return result;
    }
}

File: swift/KClosestPointstoOrigin.swift
// should use heap
class Solution {
    func kClosest(_ points: [[Int]], _ k: Int) -> [[Int]] {
        var distToPoint = [Double: [[Int]]]()
        
        for pt in points {
            let distance = sqrt(Double(pt[0] * pt[0]) + Double(pt[1] * pt[1]))
            
            if distToPoint[distance] == nil {
                distToPoint[distance] = [pt]
            } else {
                distToPoint[distance]! += [pt]
            }
        }
        
        var sortingKeys = distToPoint.keys.sorted()
        var res:[[Int]] = []
        let total = points.count
        
        for key in sortingKeys {
            for val in distToPoint[key]! {
                res += [val]
                
                if res.count >= k {
                    return res
                }
            }
        }
        
        return res
    }
}